dnshostprovider.go

```diff
@@ -40,49 +70,71 @@ func (hp *DNSHostProvider) Init(servers []string) error {
 		if err != nil {
 			return err
 		}
-		for _, addr := range addrs {
-			found = append(found, net.JoinHostPort(addr, port))
+
+		if hp.static {
+			for _, addr := range addrs {
+				found = append(found, hostPort{addr, port})
+			}
+		} else {
+			// In dynamic mode, the lookup is still performed to validate that the given ZK servers currently
+			// resolve, however these results are discarded as the addresses will be resolved dynamically when
+			// Next is called.
+			found = append(found, hostPort{host, port})
 		}
 	}
 
 	if len(found) == 0 {
-		return fmt.Errorf("No hosts found for addresses %q", servers)
+		return fmt.Errorf("zk: no hosts found for addresses %q", servers)
 	}
 
 	// Randomize the order of the servers to avoid creating hotspots
-	stringShuffle(found)
+	rand.Shuffle(len(found), func(i, j int) { found[i], found[j] = found[j], found[i] })
 
 	hp.servers = found
-	hp.curr = -1
-	hp.last = -1
+	hp.last = len(hp.servers) - 1
 
 	return nil
 }
 
-// Len returns the number of servers available
-func (hp *DNSHostProvider) Len() int {
-	hp.mu.Lock()
-	defer hp.mu.Unlock()
-	return len(hp.servers)
-}
-
-// Next returns the next server to connect to. retryStart will be true
-// if we've looped through all known servers without Connected() being
-// called.
+// Next returns the next server to connect to. retryStart should be true if this call to Next
+// exhausted the list of known servers without Connected being called. If connecting to this final
+// host fails, the connect loop will back off before invoking Next again for a fresh server.
 func (hp *DNSHostProvider) Next() (server string, retryStart bool) {
 	hp.mu.Lock()
 	defer hp.mu.Unlock()
-	hp.curr = (hp.curr + 1) % len(hp.servers)
 	retryStart = hp.curr == hp.last
-	if hp.last == -1 {
-		hp.last = 0
+
+	next := hp.servers[hp.curr]
+	if hp.static {
+		server = next.String()
+	} else {
+		addrs, err := hp.lookupHost(next.host)
+		if len(addrs) == 0 {
+			if err == nil {
+				// If for whatever reason lookupHosts returned an empty list of addresses but a nil error, use a
+				// default error
+				err = fmt.Errorf("zk: no hosts resolved by lookup for %q", next.host)
+			}
+			slog.Warn("Could not resolve ZK host", "host", next.host, "err", err)
+			//
+			server = next.String()
+		} else {
+			server = addrs[rand.Intn(len(addrs))] + ":" + next.port
+		}
 	}
-	return hp.servers[hp.curr], retryStart
+
+	hp.curr = (hp.curr + 1) % len(hp.servers)
+
+	return server, retryStart
 }
 
 // Connected notifies the HostProvider of a successful connection.
 func (hp *DNSHostProvider) Connected() {
 	hp.mu.Lock()
 	defer hp.mu.Unlock()
-	hp.last = hp.curr
+	if hp.curr == 0 {
+		hp.last = len(hp.servers) - 1
+	} else {
+		hp.last = hp.curr - 1
+	}
```

Comment:
Just a note that this will now throttle retries even if a connection was never established, which is better than the previous implementation, so it looks good. Hope this fact was the motivation here.

Reply:
Right, if there's only one host in the list, it will cause it to back off between retrying that host. That was indeed the intent. However if there's multiple hosts it won't back off in between